File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.js
/**
 * @param {number[]} prices
 * @return {number}
 */
var maxProfit = function(prices) {
    if(prices.length <= 1) return 0;
    let p1 = prices[0];
    let i=1;
    let profit = 0;

    while(i < prices.length) {
        // more profit found?
        profit = Math.max(profit, prices[i] - p1);

        // less price found?
        if(prices[i] < p1) p1 = prices[i];

        // move to next
        ++i;
    }

    return profit;
};

File: 0121-best-time-to-buy-and-sell-stock/0121-best-time-to-buy-and-sell-stock.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        i=0     # point to the lower left
        j=1     # point to the higher on right of i
        mx=0    # max profit

        while j < len(prices):
            # if this price is more than price at i,
            # then, see if this is the new max
            if prices[j] > prices[i]:
                mx = max(mx, prices[j] - prices[i])
            # else we have found a new low (lower than current i)
            # point i to this new low
            else:
                i = j
            # keep forwarding j
            j += 1
        
        return mx
